### PR TITLE
Rework all network part

### LINKED ISSUE #20: Change network discovery behavior
We shouldn't hardcode the behavior of network interface discovery and instead use parameters in the configuration file to let the user choose what they want to ignore

That way we will manage corner cases of users having loopback IP (ipvs, bgp etc)
And also create vlan interfaces, which is not working right now

I propose something like:

```
network:
  ignore_interfaces: "(dummy.*|docker.*)"
  ignore_ips: "(127\.0\.0\..*)"
```

### PR DESCRIPTION
**Pull request by Solvik**
Fixes #20 and #22 

- [x] Create interface if it doesn't match IGNORE_INTERFACES regex
- [x] Create IP if it doesn't match IGNORE_IPS regex
- [x] Create interface if tagged with VLAN
- [x] Set `LAG` on bonding interfaces
- [x] Set `Parent LAG Interface` on `Child` interfaces
- [x] Double check update on every interface's attributes
- [x] Set or update IPMI interface as `mgmt_only` interface
- [x] Get  or create VLAN interface and associate it

### CODE REVIEW

File: netbox_agent/ipmi.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    802.1q VLAN Priority    : 0
+    RMCP+ Cipher Suites     : 0,1,2,3
+    Cipher Suite Priv Max   : XuuaXXXXXXXXXXX
+                            :     X=Cipher Suite Unused
+                            :     c=CALLBACK
+                            :     u=USER
+                            :     o=OPERATOR
+                            :     a=ADMIN
+                            :     O=OEM
+    Bad Password Threshold  : Not Available
```
[the remark this answers is not in the document]
Comment (Solvik) [reply netbox_agent/ipmi.py]: doesn't work with `lan print`

File: README.md [hunk 2]
```diff
@@ -8,16 +8,25 @@ The goal is to generate an existing infrastructure on Netbox and have the abilit
...
+- Python >= 3.4
+- [python3-netaddr](https://github.com/drkjam/netaddr)
+- [python3-netifaces](https://github.com/al45tair/netifaces)
+
 # Known limitations
 
 * The project is only compatible with Linux.
 Since it uses `ethtool` and parses `/sys/` directory, it's not compatible with *BSD distributions.
 * Netbox `>=2.6.0,<=2.6.2` has a caching problem ; if the cache lifetime is too high, the script can get stale data after modification.
+We advise to set `CACHE_TIME` to `0`.
```
[the remark this answers is not in the document]
Comment (Solvik) [reply README.md]: There's a few issue on the netbox project to fix the cache invalidation upon modification
Not our case to fix

[the remark this answers is not in the document]
Comment (ramnes) [reply README.md]: jup right

File: README.md [hunk 3]
```diff
@@ -8,16 +8,25 @@ The goal is to generate an existing infrastructure on Netbox and have the abilit
...
 * Update existing `Device` and `Interfaces`
 * Handle blade moving (new slot, new chassis)
 
+# Requirements
+
+- Netbox >= 2.6
+- Python >= 3.4
+- [python3-netaddr](https://github.com/drkjam/netaddr)
+- [python3-netifaces](https://github.com/al45tair/netifaces)
+
```
[the remark this answers is not in the document]
Comment (Solvik) [reply README.md]: Good idea!

File: netbox_agent/ipmi.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        self.ret, self.output = subprocess.getstatusoutput('ipmitool lan print')
+        if not self.ret:
+            logging.error('Cannot get ipmi info: {}'.format(self.output))
+
+
+    def parse(self):
+        ret = {}
+        if self.ret != 0:
+            return ret
+        for line in self.output.split('\n'):
```
[the remark this answers is not in the document]
Comment (Solvik) [reply netbox_agent/ipmi.py]: Will fix

File: netbox_agent/network.py [hunk 5]
```diff
@@ -79,28 +143,77 @@ def get_netbox_type_for_nic(self, nic):
...
+        )
+
+    def create_or_update_ipmi(self):
+        ipmi = self.get_ipmi()
+        mac = ipmi['MAC Address']
+        ip = ipmi['IP Address']
+        netmask = ipmi['Subnet Mask']
+        vlan = ipmi['802.1q VLAN ID'] if ipmi['802.1q VLAN ID'] != 'Disabled' else None
+
+        address = IPNetwork('{}/{}'.format(ip, netmask)).__str__()
```
[the remark this answers is not in the document]
Comment (Solvik) [reply netbox_agent/network.py]: Will fix

File: netbox_agent/network.py [hunk 6]
```diff
@@ -33,40 +35,102 @@
...
-            if re.match(INTERFACE_REGEX, interface):
+            # ignore if it's not a link (ie: bonding_masters etc)
+            if not os.path.islink('/sys/class/net/{}'.format(interface)):
+                continue
+
+            if NETWORK_IGNORE_INTERFACES and \
+               re.match(NETWORK_IGNORE_INTERFACES, interface):
+                logging.debug('Ignore interface {interface}'.format(interface=interface))
+                continue
+            else:
```
[the remark this answers is not in the document]
Comment (Solvik) [reply netbox_agent/network.py]: Good catch, will fix

File: netbox_agent/ipmi.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
+import logging
+import subprocess
+
+class Ipmi():
```
[the remark this answers is not in the document]
Comment (Solvik) [reply netbox_agent/ipmi.py]: Will fix

File: netbox_agent/config.py [hunk 8]
```diff
@@ -30,3 +30,9 @@
     RACK_LOCATION_DRIVER_FILE = rack_location.get('driver_file')
     RACK_LOCATION = rack_location.get('driver')
     RACK_LOCATION_REGEX = rack_location.get('regex')
+
+NETWORK_IGNORE_INTERFACES = None
+NETWORK_IGNORE_IPS = None
+if config.get('network'):
+    NETWORK_IGNORE_INTERFACES = config['network']['ignore_interfaces']
+    NETWORK_IGNORE_IPS = config['network']['ignore_ips']
```
[the remark this answers is not in the document]
Comment (Solvik) [reply netbox_agent/config.py]: Good catch

File: netbox_agent/ipmi.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ramnes) [reply netbox_agent/ipmi.py]: tristesse

File: netbox_agent/network.py [hunk 9]
```diff
@@ -33,40 +35,102 @@
...
+                    'ethtool': Ethtool(interface).parse(),
+                    'vlan': vlan,
+                    'bonding': bonding,
+                    'bonding_slaves': bonding_slaves,
                 }
                 self.nics.append(nic)
 
+    def _set_bonding_interfaces(self):
+        logging.debug('Setting bonding interfaces..')
+        for nic in [x for x in self.nics if x['bonding']]:
```
[the remark this answers is not in the document]
Comment (Solvik) [reply netbox_agent/network.py]: I didn't even know about generator comprehension, thansk a lot !!
Will fixx

**Pull request merged**